theano/tensor/basic.py

```diff
@@ -2523,7 +2523,7 @@ def grad(self, inputs, grads):
             new_order = list(x.broadcastable)
             idx = 0
             for i in range(x.ndim):
-                if not new_order[i]:
+                if i not in axis_broadcasted:
```

Comment:
I am not sure about this solution. What if `n_axes_to_sum` is greater then zero? Then indices in `axis_broadcasted` correspond to `gz` dimensions, not `x` dimensions.

Here is how I fixed it in my fork:

```diff
+        axis_kept = []
         for i, (ib, gb) in enumerate(
             zip(inputs[0].broadcastable,
                 # We need the dimensions corresponding to x
                 grads[0].broadcastable[-inputs[0].ndim:])):
             if ib and not gb:
                 axis_broadcasted.append(i + n_axes_to_sum)
+            else:
+                axis_kept.append(i)
         gx = gz.sum(axis=axis + axis_broadcasted)
         if axis_broadcasted:
-            new_order = list(x.broadcastable)
-            idx = 0
-            for i in range(x.ndim):
-                if not new_order[i]:
-                    new_order[i] = idx
-                    idx += 1
-                else:
-                    new_order[i] = 'x'
+            new_order = ['x'] * x.ndim
+            for idx, axis in enumerate(axis_kept):
+                new_order[axis] = idx
--